dvc/external_repo.py

```diff
@@ -123,21 +123,32 @@ def _set_cache_dir(self):
 
         self.cache.local.cache_dir = cache_dir
 
-    def _set_upstream(self):
-        # check if the URL is local and no default remote is present
-        # add default remote pointing to the original repo's cache location
-        if os.path.isdir(self.url):
-            if not self.config["core"].get("remote"):
-                src_repo = Repo(self.url)
-                try:
-                    cache_dir = src_repo.cache.local.cache_dir
-                finally:
-                    src_repo.close()
-
-                self.config["remote"]["auto-generated-upstream"] = {
-                    "url": cache_dir
-                }
-                self.config["core"]["remote"] = "auto-generated-upstream"
+    def _fix_upstream(self):
+        if not os.path.isdir(self.url):
+            return
+
+        remote_name = self.config["core"].get("remote")
+        src_repo = Repo(self.url)
+        try:
+            if remote_name:
+                self._fix_local_remote(src_repo, remote_name)
+            else:
+                self._add_upstream(src_repo)
+        finally:
+            src_repo.close()
+
+    def _fix_local_remote(self, src_repo, remote_name):
+        new_remote = self.config["remote"][remote_name]
+        old_remote = src_repo.config["remote"][remote_name]
+        if new_remote["url"] != old_remote["url"]:
+            new_remote["url"] = old_remote["url"]
+
```

Comment:
Would at least add a comment about what is going on here, though.

Reply:
Done. Btw, should I squash all commits into a single one?

Reply:
As you wish. But we'll squash automatically anyways :slightly_smiling_face: Though if you think this is a final patch, you need to remove WIP prefix from the header.

Reply:
Ah cool! no need then. "WIP" removed.